Support k8s api discovery and group api

moolitayer opened the pull request:
#### Creating the client
Currently we support arguments for kubernetes api path and version and guess '/api' and 'v1' if not provided.
Both of these are valid today and create a client for the same endpoint:
```
client = Kubeclient::Client.new('http://localhost:8080/api', 'v1')
client = Kubeclient::Client.new('http://localhost:8080')
```

It is now possible to create a client for a group api:
```
client = Kubeclient::Client.new('http://localhost:8080/apis/batch', 'v1')
```
#### Discovery
Discovery is done lazily on method calls so it should not change behavior. 

It can also  be done explicitly:
```
client = Kubeclient::Client.new('http://localhost:8080/api', 'v1')
client.discover
```

It is possible to check the status of discovery
```
unless client.discovered
  client.discover
end
```

### Non Backward compatible changes
#### Class definitions

It is no longer possible to support the following code:
```
require 'kubeclient'

service = Kubeclient::Service.new
```

Since classes are no longer statically defined they will not be available at require time.

One solution for this added here is to use the generic `Resource`:
```
service = Kubeclient::Resource.new

```
used to fix some of the tests. I suggest to use this everywhere (including the Readme) since specific entity classes hold no logic and the following will work:

```
client.create_pod(Kubeclient::Service.new)
``` 

Class definitions after the discovery would work but I do not suggest to relay on that .

### Resources in need of special attention
Here is an example for how a normal resource looks in discovery:
```
    {
      "name": "componentstatuses",
      "namespaced": false,
      "kind": "ComponentStatus"
    },
```

####  Endpoints
```
    {
      "name": "endpoints",
      "namespaced": true,
      "kind": "Endpoints" # kind is plural!
    },
```
See https://github.com/kubernetes/kubernetes/issues/8115 complete backward compatibility is currently maintained with specific code in discovery and Endpoint creation. 

As a side note, it would be possible to drop the special handling (and change interface) if we follow https://github.com/kubernetes/kubernetes/issues/18622 and oc client and have kubeclient react the same to plural and singular methods. to distinguish between get one and get all we would have to look at the parameters but that seems completely doable.

#### Resources not supporting GET
```
    {
      "name": "bindings",
      "namespaced": true,
      "kind": "Binding"
    },
```
Bindings is an example for a resource that does not support get, all_entities ignores exceptions so that it keeps working. I don't think there is a way to learn that it does not support get in discovery except for queering it. 

#### Non verbs
Some resources do not fall under the logic kubeclient currently support and should be ignored by discovery. These include:
From k8s /apis/extensions/v1beta1

```
    {
      "name": "replicationcontrollers",
      "namespaced": true,
      "kind": "ReplicationControllerDummy"
    },
```

 See also: https://github.com/kubernetes/kubernetes/issues/22413

OpenShift:
 
```
    {
      "name": "generatedeploymentconfigs",
      "namespaced": true,
      "kind": "DeploymentConfig"
    },
...
    {
      "name": "processedtemplates",
      "namespaced": true,
      "kind": "Template"
    },
```

As reference, all the resource currently returned in discovery from k8s core and groups as well as OpenShift can be found [here](https://gist.github.com/moolitayer/c5d831936bef690aa641089e3a55743a)

### openshift_client
Since OpenShift support discovery in the same way, It is possible to create a client for OpenShift using Kubeclient:
```
Kubeclient::Client.new('http://localhost:8443/oapi')
```

So it could be possible in the future to drop openshift_client in favor of kubeclient if that is desirable.

This also brings the possibility of having a forwarding client - just like how oc can work on k8s entities. That would be helpful for some current efforts like generation from templates but is outside the scope of this pr.

### Out of the scope of this pr
- Discovery of nested resources (```namespaces/finalize```)
- Discovery of first level resources (```/healthz```)

This continues the work done in #184 to drop activesupport removing the last method that had lingual logic: pluralize_entities in favor of discovery..

Fixes #144  
Fixes #189

moolitayer commented:
@abonas @simon3z please review

simon3z commented:
@jonmoter @jimmidyson @nyxcharon can you review?

simon3z commented:
cc @Fryguy

simon3z commented:
> After a client is created and before it can be used discovery must be made.

@moolitayer can't you just hide that with a lazy initialization?

jonmoter commented:
We have a lot of logic in our application that does the equivalent of:

```
require 'kubeclient'
service = Kubeclient::Service.new
```

and then we have unit tests that exercise that functionality. I don't want to require a Kubernetes server to be running in order to run my unit tests.

I see that you have some logic to stub out the discovery call and return the contents of `core_api_resource_list.json`.  Could you wrap that logic up in a helper method that could be used when running tests?  Maybe something like:

```
require 'kubeclient'
require 'kubeclient/test'

Kubeclient.stub_core_apis!
# rest of my tests
```

That would make it far easier to cope with the breaking changes this introduces.

/cc @grosser, @irwaters

moolitayer commented:
> @moolitayer can't you just hide that with a lazy initialization?

now supporting lazy initialization

moolitayer commented:
@jonmoter could you elaborate on your tests:
are you doing something with the service? e.g 
```
client.create_service(service)
```
if so how are you mocking that?

if not will using 
```
service = Kubeclient::Resource.new
```
solve your problem?

jonmoter commented:
In our model code, we do stuff like:
```
client.create_service(Kubeclient::Service.new(service_template))
```
and then in our tests, we stub out `create_service` to return some value.

So now that I think about it, we could change our model logic to say `Kubeclient::Resource.new(service_template)`, which would be a minimal change.  That's not too bad.

simon3z commented on lib/kubeclient/common.rb:
```diff
@@ -51,6 +53,9 @@ def initialize_client(
       handle_uri(uri, path)
 
+      @class_owner = class_owner
+      @entities = nil
```
@moolitayer why setting this to `nil` when the first thing you do in `discover` is to set it to an empty hash?

moolitayer replied:
The only goal here is to have all instance variables explicitly mentioned at initialization - it serves no functional purpose. 

cc @abonas

simon3z replied:
@moolitayer I wasn't suggesting to remove it, I was suggesting to use the initialization value that you need (empty hash).

moolitayer replied:
That I can do. I would like to keep the same initialization in discover  - so that it purges the list of entities each time it is called.

simon3z commented on lib/kubeclient/common.rb:
```diff
@@ -51,6 +53,9 @@ def initialize_client(
 
+      @class_owner = class_owner
+      @entities = nil
+      @discovered = false
```
@moolitayer isn't `@discovered` the same as `@entities.empty?`? (provided that you initialize `@entities` to an hash). If you really want to keep it explicit then you can have a method:

```ruby
    def discovered
      @entities.empty?
    end
```

moolitayer replied:
I noticed in case the returned APIResourceList is empty or has only things that aren't parsed (like nested resources) that could cause an infinite loop in method_missing. 
I would like to have some guard against that

simon3z commented on lib/kubeclient/common.rb:
```diff
@@ -78,11 +92,48 @@ def handle_exception
+          @entities[entity.method_names[0]] = entity if entity
+        end
+        define_entity_methods
+        @discovered = true
```
@moolitayer why all these lines are within the `handle_exception` block? Shouldn't just the `rest_client` request be within the `handle_exception` block?

simon3z commented on lib/kubeclient/common.rb:
```diff
@@ -78,11 +92,48 @@ def handle_exception
+      # "kind": "ComponentStatus" NetworkPolicy, Endpoints
+      # maintain pre group api compatibility for endpoints.
+      # See: https://github.com/kubernetes/kubernetes/issues/8115
+      kind = 'Endpoint' if kind == 'Endpoints'
```
@moolitayer this looks like another exception as the one you have at line 102 (`ReplicationControllerDummy`). I suggest to group these normalizations in a single method you can reuse.

moolitayer replied:
Forgot to remove the exception for ReplicationControllerDummy after implementing a general solution for non verbs.

moolitayer commented:
Addressed comments @simon3z @abonas PTAL

simon3z commented:
@jonmoter @jimmidyson @nyxcharon @Fryguy @abonas @zeari @yaacov @cben any comment on this?
We'd like to merge this ASAP (btw we already looked at this in another fashion in #181)

jimmidyson commented:
Sorry not got time to review right now - I'm sure what you add will be great!

moolitayer commented:
- Updated README
- Addressed comment about ```@entities``` initialization

simon3z commented on lib/kubeclient/common.rb:
```diff
@@ -78,11 +92,47 @@ def handle_exception
+      @entities = {}
+      response = handle_exception do
+        rest_client.get(@headers)
+      end
```
@moolitayer will a one-liner be nicer here?

```ruby
response = handle_exception { rest_client.get(@headers) }
```

Same for other similar instances in the rest of the PR.

simon3z replied:
@moolitayer maybe even (if not too clever):

```ruby
result = JSON.parse(handle_exception { rest_client.get(@headers) })
```

moolitayer replied:
That bugged me too.
It is done elsewhere so I submitted:
https://github.com/abonas/kubeclient/pull/186
(For the other one line handler)

and I will fix this one here

simon3z commented on lib/kubeclient/common.rb:
```diff
@@ -66,6 +71,15 @@ def initialize_client(
       end
     end
 
+    def method_missing(method_sym, *args, &block)
```
@moolitayer I don't think the `super` part is needed. I have the feeling that this logic is simpler and may work as well:

```ruby
def method_missing(method_sym, *args, &block)
  discover unless !@discovered
  send(method_sym, *args, &block)
end
```

moolitayer replied:
Causes an infinite loop since method_missing will keep getting called

Fryguy replied:
Always define `respond_to_missing?` when implementing `method_missing`... https://robots.thoughtbot.com/always-define-respond-to-missing-when-overriding

simon3z commented on lib/kubeclient/common.rb:
```diff
@@ -51,6 +53,9 @@ def initialize_client(
 
+      @class_owner = class_owner
+      @entities = {}
+      @discovered = false
```
@moolitayer if you wanted to initialize all the attributes you are missing `@group` I think.

moolitayer replied:
That is done in handle uri (same like @api_endpoint) so I now always init it there.
also renamed group to api_group (like existing api_version) and changed code to rely on initialization.

moolitayer commented:
@simon3z PTAL

zeari commented:
@moolitayer I think this looks good but im not too knowledgeable in this code.

enoodle commented:
:+1: LGTM, I will try to use it now to add deamonset and hopfully I will not have any problems with it.

enoodle commented on lib/kubeclient.rb:
```diff
@@ -12,40 +12,20 @@ module Kubeclient
       initialize_client(
+        Kubeclient,
         uri,
         '/api',
```
@moolitayer do you think its possible to make this parameter customizable by the user as well? Its needed for example to access beta api: http://kubernetes.io/docs/api/ ( /apis/extentions/<version> )

moolitayer replied:
The user can change this (please give it a try), '/api' is only the default if a user used a uri that includes no path:
```client = Kubeclient::Client.new('http://localhost:8080')```

enoodle replied:
:+1: now I see it, Thanks!

simon3z commented on lib/kubeclient/common.rb:
```diff
@@ -66,6 +71,15 @@ def initialize_client(
     end
 
+    def method_missing(method_sym, *args, &block)
+      if !@discovered && method_sym.to_s.match(/^(get|watch|delete|create|update|patch)_(.*)$/)
```
@moolitayer do we need to check for these or is it just an optimization that we could drop? Can we make a constant and simplify the check?

```ruby
ENTITY_METHODS = %w(get watch delete create update patch)

ENTITY_METHODS.any? { |x| method_sym.to_s.start_with?(x) }
```

I am not sure if I like the above better, anyway what I would like to have is a single explicit place where we can add these verbs.

moolitayer replied:
> @moolitayer do we need to check for these or is it just an optimization that we could drop?

Since this is using a basic mechanism of the language I prefer to be as explicit as possible to avoid replacing any unexpected methods. I will extract the list as suggested.

simon3z commented:
> client = Kubeclient::Client.new('http://localhost:8080/api', 'v1')

@moolitayer we had a way to discover the available versions. Is that still available?

simon3z commented:
@moolitayer IIRC @zakiva may need a new verb `process`. See #183. Any chance to squeeze that in (another patch but this very PR). That would probably help him.

simon3z commented on lib/kubeclient/common.rb:
```diff
@@ -91,52 +138,60 @@ def build_namespace_prefix(namespace)
+    end
 
+    def define_entity_methods
+      @entities.values.each do |entity|
```
@moolitayer once you'll have extracted the verbs (also including the missing `process`), please file an issue (or a trello card) to remember that maybe we can cleanup this method looping over the verbs.

moolitayer replied:
you mean with the list of verbs? each method has different parameters set so I'm not sure how that would work

enoodle commented:
@simon3z I think that `process` only applies for templates so it should be handled separately.

simon3z commented:
> @simon3z I think that process only applies for templates so it should be handled separately.

@enoodle separate how? code? patch? PR?

enoodle commented:
@simon3z I thought it would be defined in openshift_client as a sole method `process_template` that will handle this specific case.

moolitayer commented:
> @moolitayer IIRC @zakiva may need a new verb process. See #183. Any chance to squeeze that in (another patch but this very PR). That would probably help him.

Any reason not to keep it in a separate patch?

simon3z commented:
> > @moolitayer IIRC @zakiva may need a new verb process. See #183. Any chance to squeeze that in (another patch but this very PR). That would probably help him.
>
> Any reason not to keep it in a separate patch?

@moolitayer, the argument was that `process` for now is just an OpenShift verb. Although there's a proposal here for Kubernetes https://github.com/kubernetes/kubernetes/pull/25293

@enoodle @moolitayer if this PR is covering OpenShift as a group API I feel that having another gem just for the `process` verb is overkill, and anyway the Kubernetes PR will be merged at some point.

simon3z commented:
LGTM :+1:
@abonas any opinion on this PR?

cc @jonmoter @nyxcharon @Fryguy @zeari @yaacov @cben @grosser @irwaters

moolitayer commented:
Addressed with the best solution I found. Added a test test_missing_method.rb because this is quite hairy. 
- I've reset the client in the test since it depends on state and I noticed some of the implementations I tested were wrong only on the first call.
- Currently respond_to? will raise an exception if the discovery fails. We could also return false. I'm not sure what is more deceiving for a user...

@Fryguy thanks, please take another look.

Fryguy commented:
@moolitayer The respond_to_missing? changes look great to me :+1:

zeari commented:
:hooray:

zeari commented:
:confetti_ball:
